cmd/storagenode-updater/main_test.go

```diff
@@ -4,68 +4,194 @@
 package main_test
 
 import (
+	"archive/zip"
+	"fmt"
+	"io"
 	"io/ioutil"
 	"net/http"
 	"net/http/httptest"
 	"os"
 	"os/exec"
+	"path/filepath"
+	"runtime"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"go.uber.org/zap/zaptest"
 
 	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/internal/testidentity"
+	"storj.io/storj/internal/version"
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/storj"
 	"storj.io/storj/versioncontrol"
 )
 
-func TestAutoUpdater(t *testing.T) {
+const (
+	oldVersion = "v0.19.0"
+	newVersion = "v0.19.5"
+)
+
+func TestAutoUpdater_unix(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("requires storagenode and storagenode-updater to be installed as windows services")
+	}
+
+	// TODO cleanup `.exe` extension for different OS
+
 	ctx := testcontext.New(t)
 	defer ctx.Cleanup()
 
-	testFiles := []struct {
-		src   string
-		dst   string
-		perms os.FileMode
-	}{
-		{
-			"testdata/fake-storagenode",
-			ctx.File("storagenode"),
-			0755,
-		},
-		{
-			"testdata/fake-ident.cert",
-			ctx.File("identity.cert"),
-			0644,
-		},
-		{
-			"testdata/fake-ident.key",
-			ctx.File("identity.key"),
-			0600,
-		},
+	oldSemVer, err := version.NewSemVer(oldVersion)
+	require.NoError(t, err)
+
+	newSemVer, err := version.NewSemVer(newVersion)
+	require.NoError(t, err)
+
+	oldInfo := version.Info{
+		Timestamp:  time.Now(),
+		CommitHash: "",
+		Version:    oldSemVer,
+		Release:    false,
 	}
 
-	for _, file := range testFiles {
-		content, err := ioutil.ReadFile(file.src)
-		require.NoError(t, err)
+	// build real storagenode and updater with old version
+	oldStoragenodeBin := ctx.CompileWithVersion("storj.io/storj/cmd/storagenode/", oldInfo)
+	storagenodePath := ctx.File("fake", "storagenode.exe")
+	copyBin(ctx, t, storagenodePath, oldStoragenodeBin)
```

Comment:
Can't we compile the binary at the desired location instead of doing file copy afterwards?

Reply:
I don't know. We can try to do this later. I can move files instead of copy for better performance.

Reply:
It may be possible. Compile create a temp directory with the name of the test, so it should be possible to rename it and used.

Reply:
Yeah, moving also sounds better then copying.